### ISSUE TITLE
Clarification needed for mime type handling

### ISSUE THREAD

**Issue by Pehrsons** [w3c/mediacapture-record#170]
Let's start by looking at the `MediaRecorderOptions` `mimeType` attribute, where the spec says:
> The container and codec format(s) [RFC2046] for the recording, which may include any parameters that are defined for the format. If the UA does not support the format or any of the parameters specified, it MUST throw a `NotSupportedError` `DOMException`. If this paramater is not specified, the UA will use a platform-specific default format. The container format, whether passed in to the constructor or defaulted, will be used as the value of the `mimeType` attribute.

This makes it sound like the `MediaRecorder`, upon running the constructor, will lock its `mimeType` attribute to either the passed in value, or a defaulted value (if invalid it throws, so all these are valid and supported).

Then if we look at the `MediaRecorder` `mimeType` attribute, the spec says:
> The MIME type [RFC2046] that has been selected as the container for recording. This entry includes all the parameters to the base mimeType. The UA should be able to play back any of the MIME types it supports for recording. For example, it should be able to display a video recording in the HTML <video> tag. The default value for this property is platform-specific. 

This also sounds like it's read-only and never changes, but it seems to go against the language for the `MediaRecorderOptions` `mimeType` attribute in that it says "This entry includes all the parameters to the base mimeType". I take it "all the parameters" (definition?) are meant to be added to whatever was passed in with `MediaRecorderOptions`?
The contradiction lies in that the language for `MediaRecorderOptions.mimeType` says "The container format, whether passed in to the constructor or defaulted, will be used as the value of the `mimeType` attribute".

Clarifying this would be Part 1 of this issue.

Part 2 is minor, but "all the parameters" in the paragraph above would be good to clarify.
- Is it a MUST to define all codecs to be used before tracks are known? Are there other parameters that MUST be listed explicitly? "all" is very broad.

Part 3 is to clarify what we are allowed to encode once the `mimeType` has been locked onto.
- Can an audio mime type (e.g., `audio/webm;codecs=opus`) encode video?
- Can a video mime type (e.g., `video/webm;codecs=vp8`) encode audio?
- Can an audio/video mime type (e.g., `video/webm;codecs=opus,vp8`) encode only audio?
- Can an audio mime type declare a video codec and encode audio and video?

To gather some information on existing behavior I wrote a fiddle to test this, https://jsfiddle.net/pehrsons/h4rwbn1e/.
It gives per the below,

Legend:
- `ctor-mime` is the recorder's `mimeType` attribute just after running the MediaRecorder constructor
- `recording-mime` is the recorder's `mimeType` attribute just after the first `dataavailable` event (timeslice is provided)
- `blob-mime` is the `type` attribute of the `Blob` in the first `dataavailable` event
- `recorded tracks` is a list of the tracks (and their kinds) that were the result of decoding the recording, i.e., to see if the `mimeType` attribute was lying

Firefox (though note that this is the behavior I'm intending to make spec compliant):
```
[MediaRecorder "" [audio]]: ctor-mime: ""; recording-mime: "audio/ogg"; blob-mime: "audio/ogg"; recorded tracks: [audio]
[MediaRecorder "" [video]]: ctor-mime: ""; recording-mime: "video/webm"; blob-mime: "video/webm"; recorded tracks: [video]
[MediaRecorder "" [audio,video]]: ctor-mime: ""; recording-mime: "video/webm"; blob-mime: "video/webm"; recorded tracks: [audio,video]
[MediaRecorder "audio/webm;codecs=opus" [audio]]: ctor-mime: "audio/webm;codecs=opus"; recording-mime: "audio/ogg"; blob-mime: "audio/ogg"; recorded tracks: [audio]
[MediaRecorder "audio/webm;codecs=opus" [video]]: ctor-mime: "audio/webm;codecs=opus"; recording-mime: "video/webm"; blob-mime: "video/webm"; recorded tracks: [video]
[MediaRecorder "audio/webm;codecs=opus" [audio,video]]: ctor-mime: "audio/webm;codecs=opus"; recording-mime: "video/webm"; blob-mime: "video/webm"; recorded tracks: [audio,video]
[MediaRecorder "video/webm;codecs=vp8" [audio]]: ctor-mime: "video/webm;codecs=vp8"; recording-mime: "audio/ogg"; blob-mime: "audio/ogg"; recorded tracks: [audio]
[MediaRecorder "video/webm;codecs=vp8" [video]]: ctor-mime: "video/webm;codecs=vp8"; recording-mime: "video/webm"; blob-mime: "video/webm"; recorded tracks: [video]
[MediaRecorder "video/webm;codecs=vp8" [audio,video]]: ctor-mime: "video/webm;codecs=vp8"; recording-mime: "video/webm"; blob-mime: "video/webm"; recorded tracks: [audio,video]
[MediaRecorder "video/webm;codecs=opus,vp8" [audio]]: ctor-mime: "video/webm;codecs=opus,vp8"; recording-mime: "audio/ogg"; blob-mime: "audio/ogg"; recorded tracks: [audio]
[MediaRecorder "video/webm;codecs=opus,vp8" [video]]: ctor-mime: "video/webm;codecs=opus,vp8"; recording-mime: "video/webm"; blob-mime: "video/webm"; recorded tracks: [video]
[MediaRecorder "video/webm;codecs=opus,vp8" [audio,video]]: ctor-mime: "video/webm;codecs=opus,vp8"; recording-mime: "video/webm"; blob-mime: "video/webm"; recorded tracks: [audio,video]
[MediaRecorder "audio/webm;codecs=opus,vp8" [audio]]: Exception; NotSupportedError: Operation is not supported
[MediaRecorder "audio/webm;codecs=opus,vp8" [video]]: Exception; NotSupportedError: Operation is not supported
[MediaRecorder "audio/webm;codecs=opus,vp8" [audio,video]]: Exception; NotSupportedError: Operation is not supported
```

Chrome:
```
[MediaRecorder "" [audio]]: ctor-mime: "audio/webm;codecs=opus"; recording-mime: "audio/webm;codecs=opus"; blob-mime: "audio/webm;codecs=opus"; recorded tracks: [audio]
[MediaRecorder "" [video]]: ctor-mime: "video/webm;codecs=vp8"; recording-mime: "video/webm;codecs=vp8"; blob-mime: "video/webm;codecs=vp8"; recorded tracks: [video]
[MediaRecorder "" [audio,video]]: ctor-mime: "video/webm;codecs=vp8,opus"; recording-mime: "video/webm;codecs=vp8,opus"; blob-mime: "video/webm;codecs=vp8,opus"; recorded tracks: [audio,video]
[MediaRecorder "audio/webm;codecs=opus" [audio]]: ctor-mime: "audio/webm;codecs=opus"; recording-mime: "audio/webm;codecs=opus"; blob-mime: "audio/webm;codecs=opus"; recorded tracks: [audio]
[MediaRecorder "audio/webm;codecs=opus" [video]]: ctor-mime: "audio/webm;codecs=opus"; recording-mime: "audio/webm;codecs=opus"; blob-mime: "audio/webm;codecs=opus"; recorded tracks: [video]
[MediaRecorder "audio/webm;codecs=opus" [audio,video]]: ctor-mime: "audio/webm;codecs=opus"; recording-mime: "audio/webm;codecs=opus"; blob-mime: "audio/webm;codecs=opus"; recorded tracks: [audio,video]
[MediaRecorder "video/webm;codecs=vp8" [audio]]: ctor-mime: "video/webm;codecs=vp8"; recording-mime: "video/webm;codecs=vp8"; blob-mime: "video/webm;codecs=vp8"; recorded tracks: [audio]
[MediaRecorder "video/webm;codecs=vp8" [video]]: ctor-mime: "video/webm;codecs=vp8"; recording-mime: "video/webm;codecs=vp8"; blob-mime: "video/webm;codecs=vp8"; recorded tracks: [video]
[MediaRecorder "video/webm;codecs=vp8" [audio,video]]: ctor-mime: "video/webm;codecs=vp8"; recording-mime: "video/webm;codecs=vp8"; blob-mime: "video/webm;codecs=vp8"; recorded tracks: [audio,video]
[MediaRecorder "video/webm;codecs=opus,vp8" [audio]]: ctor-mime: "video/webm;codecs=opus,vp8"; recording-mime: "video/webm;codecs=opus,vp8"; blob-mime: "video/webm;codecs=opus,vp8"; recorded tracks: [audio]
[MediaRecorder "video/webm;codecs=opus,vp8" [video]]: ctor-mime: "video/webm;codecs=opus,vp8"; recording-mime: "video/webm;codecs=opus,vp8"; blob-mime: "video/webm;codecs=opus,vp8"; recorded tracks: [video]
[MediaRecorder "video/webm;codecs=opus,vp8" [audio,video]]: ctor-mime: "video/webm;codecs=opus,vp8"; recording-mime: "video/webm;codecs=opus,vp8"; blob-mime: "video/webm;codecs=opus,vp8"; recorded tracks: [audio,video]
[MediaRecorder "audio/webm;codecs=opus,vp8" [audio]]: Exception; NotSupportedError: Failed to construct 'MediaRecorder': Failed to initialize native MediaRecorder the type provided (audio/webm;codecs=opus,vp8) is not supported.
[MediaRecorder "audio/webm;codecs=opus,vp8" [video]]: Exception; NotSupportedError: Failed to construct 'MediaRecorder': Failed to initialize native MediaRecorder the type provided (audio/webm;codecs=opus,vp8) is not supported.
[MediaRecorder "audio/webm;codecs=opus,vp8" [audio,video]]: Exception; NotSupportedError: Failed to construct 'MediaRecorder': Failed to initialize native MediaRecorder the type provided (audio/webm;codecs=opus,vp8) is not supported.
```

Safari: Not implemented

Given this, I think we need Part 4: Should the `MediaRecorder` `mimeType` attribute really be read-only and never change? While I do agree this makes for a simpler, less confusing, API, no implementation works this way today.

---

To recap:

1: Can the `MediaRecorder` constructor modify the `MediaRecorderOptions` `mimeType` attribute as it sets its own `mimeType` attribute?
2: Define "all the parameters" for a mime type, or rephrase. In particular, is it a MUST to include a final list of codecs?
3: What restrictions on tracks does the selected `mimeType` impose?
4: Can the `MediaRecorder` `mimeType` attribute change during the `MediaRecorder`s lifetime (to maintain behavior of current implementations)?

**Comment by jan-ivar** [comment]
> ctor-mime is the recorder's mimeType attribute just after running the MediaRecorder constructor

Right, the name threw me, but the legend helps (it is *not* the value passed into the constructor).

My 2 cents:

1. Modify the type? I'd say no. It should throw if it can't handle the input. But if parameters are absent, then see 2.
2. From "all parameters", I'd say yes, modulo "final" (see 3). E.g. pass in `"audio/webm"`, read `"audio/webm;codecs=opus"`. But neither of the browsers do this. Seems useful though.
3. `read-only` to me doesn't mean a value is constant, only that it has no setter. So in the case where no value was passed in by the app, I don't see a contradiction. if something is passed in however, I agree it seems unclear what to do. See 4.
4. I'm happy to cement existing implementation behavior here. We've taken that approach before in this spec, and it seems reasonable here as well.

If this needs clarification, I'm happy to review a PR.

**Comment by Pehrsons** [comment]
I also made [a slightly different fiddle](https://jsfiddle.net/pehrsons/bmfuo2yr/). This one creates the `MediaRecorder` with an empty `MediaStream`, but adds the tracks before `start()`, to show some of the fundamental problems of setting the mime type and the tracks at different times.

Firefox:
```
[MediaRecorder "" [audio]]: ctor-mime: ""; recording-mime: "audio/ogg"; blob-mime: "audio/ogg"; recorded tracks: [audio]
[MediaRecorder "" [video]]: ctor-mime: ""; recording-mime: "video/webm"; blob-mime: "video/webm"; recorded tracks: [video]
[MediaRecorder "" [audio,video]]: ctor-mime: ""; recording-mime: "video/webm"; blob-mime: "video/webm"; recorded tracks: [audio,video]
[MediaRecorder "audio/webm;codecs=opus" [audio]]: ctor-mime: "audio/webm;codecs=opus"; recording-mime: "audio/ogg"; blob-mime: "audio/ogg"; recorded tracks: [audio]
[MediaRecorder "audio/webm;codecs=opus" [video]]: ctor-mime: "audio/webm;codecs=opus"; recording-mime: "video/webm"; blob-mime: "video/webm"; recorded tracks: [video]
[MediaRecorder "audio/webm;codecs=opus" [audio,video]]: ctor-mime: "audio/webm;codecs=opus"; recording-mime: "video/webm"; blob-mime: "video/webm"; recorded tracks: [audio,video]
[MediaRecorder "video/webm;codecs=vp8" [audio]]: ctor-mime: "video/webm;codecs=vp8"; recording-mime: "audio/ogg"; blob-mime: "audio/ogg"; recorded tracks: [audio]
[MediaRecorder "video/webm;codecs=vp8" [video]]: ctor-mime: "video/webm;codecs=vp8"; recording-mime: "video/webm"; blob-mime: "video/webm"; recorded tracks: [video]
[MediaRecorder "video/webm;codecs=vp8" [audio,video]]: ctor-mime: "video/webm;codecs=vp8"; recording-mime: "video/webm"; blob-mime: "video/webm"; recorded tracks: [audio,video]
[MediaRecorder "video/webm;codecs=opus,vp8" [audio]]: ctor-mime: "video/webm;codecs=opus,vp8"; recording-mime: "audio/ogg"; blob-mime: "audio/ogg"; recorded tracks: [audio]
[MediaRecorder "video/webm;codecs=opus,vp8" [video]]: ctor-mime: "video/webm;codecs=opus,vp8"; recording-mime: "video/webm"; blob-mime: "video/webm"; recorded tracks: [video]
[MediaRecorder "video/webm;codecs=opus,vp8" [audio,video]]: ctor-mime: "video/webm;codecs=opus,vp8"; recording-mime: "video/webm"; blob-mime: "video/webm"; recorded tracks: [audio,video]
[MediaRecorder "audio/webm;codecs=opus,vp8" [audio]]: Exception; NotSupportedError: Operation is not supported
[MediaRecorder "audio/webm;codecs=opus,vp8" [video]]: Exception; NotSupportedError: Operation is not supported
[MediaRecorder "audio/webm;codecs=opus,vp8" [audio,video]]: Exception; NotSupportedError: Operation is not supported
```

Chrome:
```
[MediaRecorder "" [audio]]: ctor-mime: ""; recording-mime: ""; blob-mime: ""; recorded tracks: [audio]
[MediaRecorder "" [video]]: ctor-mime: ""; recording-mime: ""; blob-mime: ""; recorded tracks: [video]
[MediaRecorder "" [audio,video]]: ctor-mime: ""; recording-mime: ""; blob-mime: ""; recorded tracks: [audio,video]
[MediaRecorder "audio/webm;codecs=opus" [audio]]: ctor-mime: "audio/webm;codecs=opus"; recording-mime: "audio/webm;codecs=opus"; blob-mime: "audio/webm;codecs=opus"; recorded tracks: [audio]
[MediaRecorder "audio/webm;codecs=opus" [video]]: ctor-mime: "audio/webm;codecs=opus"; recording-mime: "audio/webm;codecs=opus"; blob-mime: "audio/webm;codecs=opus"; recorded tracks: [video]
[MediaRecorder "audio/webm;codecs=opus" [audio,video]]: ctor-mime: "audio/webm;codecs=opus"; recording-mime: "audio/webm;codecs=opus"; blob-mime: "audio/webm;codecs=opus"; recorded tracks: [audio,video]
[MediaRecorder "video/webm;codecs=vp8" [audio]]: ctor-mime: "video/webm;codecs=vp8"; recording-mime: "video/webm;codecs=vp8"; blob-mime: "video/webm;codecs=vp8"; recorded tracks: [audio]
[MediaRecorder "video/webm;codecs=vp8" [video]]: ctor-mime: "video/webm;codecs=vp8"; recording-mime: "video/webm;codecs=vp8"; blob-mime: "video/webm;codecs=vp8"; recorded tracks: [video]
[MediaRecorder "video/webm;codecs=vp8" [audio,video]]: ctor-mime: "video/webm;codecs=vp8"; recording-mime: "video/webm;codecs=vp8"; blob-mime: "video/webm;codecs=vp8"; recorded tracks: [audio,video]
[MediaRecorder "video/webm;codecs=opus,vp8" [audio]]: ctor-mime: "video/webm;codecs=opus,vp8"; recording-mime: "video/webm;codecs=opus,vp8"; blob-mime: "video/webm;codecs=opus,vp8"; recorded tracks: [audio]
[MediaRecorder "video/webm;codecs=opus,vp8" [video]]: ctor-mime: "video/webm;codecs=opus,vp8"; recording-mime: "video/webm;codecs=opus,vp8"; blob-mime: "video/webm;codecs=opus,vp8"; recorded tracks: [video]
[MediaRecorder "video/webm;codecs=opus,vp8" [audio,video]]: ctor-mime: "video/webm;codecs=opus,vp8"; recording-mime: "video/webm;codecs=opus,vp8"; blob-mime: "video/webm;codecs=opus,vp8"; recorded tracks: [audio,video]
[MediaRecorder "audio/webm;codecs=opus,vp8" [audio]]: Exception; NotSupportedError: Failed to construct 'MediaRecorder': Failed to initialize native MediaRecorder the type provided (audio/webm;codecs=opus,vp8) is not supported.
[MediaRecorder "audio/webm;codecs=opus,vp8" [video]]: Exception; NotSupportedError: Failed to construct 'MediaRecorder': Failed to initialize native MediaRecorder the type provided (audio/webm;codecs=opus,vp8) is not supported.
[MediaRecorder "audio/webm;codecs=opus,vp8" [audio,video]]: Exception; NotSupportedError: Failed to construct 'MediaRecorder': Failed to initialize native MediaRecorder the type provided (audio/webm;codecs=opus,vp8) is not supported.
```

**Comment by Pehrsons** [comment]
@yellowdoge your thoughts?

**Comment by Pehrsons** [comment]
I'm leaning towards a proposal that cements current behavior, but makes the `mimeType` attribute accurate while we're recording.

Basically, allow the UA to extend on the mime type that was passed in to the constructor, but not modify it. This would happen in `start()` where tracks are known, and be restored to the mime type passed in to the constructor when recording stops.

This would mean that an audio mime type is not allowed to record a video track. A passed-in audio codec is locked to be used for all audio tracks, but if an audio codec is not passed in we reflect the one used for encoding in the mimetype after start(). Similarly for video. If a container is specified we'd also restrict the tracks to that, in case it, or the UA, doesn't support a track type present in the stream on `start()` we'd throw.

This clarification would have the side effect of throwing in `start()` in case the tracks present in the stream are violating the passed-in mime type.

Any immediate thoughts on this? @jan-ivar? If not I'll write up a PR.

**Comment by guest271314** [comment]
@Pehrsons 

> A passed-in audio codec is locked to be used for all audio tracks

If a change is to be made would suggest considering the path that `MediaSource` relatively recently specified at `changeType()` (https://rawgit.com/wicg/media-source/codec-switching/index.html; https://googlechrome.github.io/samples/media/sourcebuffer-changetype.html; https://github.com/web-platform-tests/wpt/pull/17416; https://github.com/web-platform-tests/wpt/pull/11618; see also https://github.com/w3c/media-source/issues/155).

Given that multiple input video and audio tracks can ultimately be recorded by a single `MediaRecorder` instance, the `MediaRecorder` instance should have the ability to change codec type during the recording, for example, from `video/x-matroska;codecs=avc1` to `video/webm;codecs=vp9`, etc.

**Comment by Pehrsons** [comment]
> @Pehrsons
> 
> > A passed-in audio codec is locked to be used for all audio tracks
> 
> If a change is to be made would suggest considering the path that `MediaSource` relatively recently specified at `changeType()` (https://rawgit.com/wicg/media-source/codec-switching/index.html; https://googlechrome.github.io/samples/media/sourcebuffer-changetype.html; [web-platform-tests/wpt#17416](https://github.com/web-platform-tests/wpt/pull/17416); [web-platform-tests/wpt#11618](https://github.com/web-platform-tests/wpt/pull/11618); see also [w3c/media-source#155](https://github.com/w3c/media-source/issues/155)).
> 
> Given that multiple input video and audio tracks can ultimately be recorded by a single `MediaRecorder` instance, the `MediaRecorder` instance should have the ability to change codec type during the recording, for example, from `video/x-matroska;codecs=avc1` to `video/webm;codecs=vp9`, etc.

If the spec is ever changed so that a single recording session can issue more than one self-contained file, then this sounds reasonable. For now, this is out of scope.

**Comment by guest271314** [comment]
@Pehrsons 

> If the spec is ever changed so that a single recording session can issue more than one self-contained file, then this sounds reasonable. For now, this is out of scope.

Where does the specification state that only a single codec has to be used? 

Are you sure a Matroska container cannot handle multiple video tracks having different codecs?

**Comment by Pehrsons** [comment]
The MediaRecorder doesn't have an API that lets you set the codec per-track. It's dependent on the mime type which contains an unordered list of codecs. Also, the MediaRecorder is not based on the matroska container; it essentially needs to handle all containers that UAs might want to add support for.

Feel free to file a separate issue if this is functionality you need, but for this issue it is out of scope.
